src/file_utils.R

```diff
@@ -170,6 +182,52 @@ zip_meteo_groups <- function(outfile, xwalk_meteo_fl_names, grouped_meteo_fls){
   scipiper::sc_indicate(outfile, data_file = data_files)
 }
 
+filter_resample_obs <- function(outfile, obs_feather, site_ids, obs_start, obs_stop, sample_res = 0.5){
+  
+  feather::read_feather(obs_feather) %>%
+    filter(site_id %in% site_ids) %>%
+    filter(date >= obs_start & date <= obs_stop) %>%
+    group_by(date, depth, site_id, source) %>%
+    summarize(temp = mean(temp)) %>%
+    ungroup() %>%
+    # interpolate
+    group_by(site_id, date, source) %>%
+    do({
+      date_df <- .
+      date_df %>%
+        mutate(
+          # find the closest PGDL depth to the observation depth
+          new_depth = purrr::map_dbl(depth, function(obsdep) {
+            # get depths from 0 to max depth of obs, rounded up to nearest sample_res, but not beyond
+            # so if obsdep = c(0.3, 9.6), we'll go to 10m, but if obsdep = c(0.3, 9.5), we'd go to 9.5
+            depths <- seq(0, max(ceiling(obsdep / sample_res) * sample_res), by = sample_res)
+            depths[which.min(abs(obsdep - depths))]
+            }),
+          # estimate temperature at the new depth using interpolation, or if
+          # that's not possible, set to the nearest observed temperature
+          new_temp = if(nrow(date_df) >= 2) {
+            # if we have enough values on this date, interpolate
+            approx(x=depth, y=temp, xout=new_depth, rule=2)$y
+          } else {
+            # if we can't interpolate, just use the nearest value
+            temp
+          },
+          depth_diff = abs(depth - new_depth)) %>%
+        # after approx(), trash any values at new_depth >= 0.5 m from the nearest observation
+        filter(depth_diff < 0.5) %>%
+        # only keep one estimate for each new_depth
+        group_by(new_depth) %>%
+        filter(depth_diff == min(depth_diff)) %>%
```

Comment:
Is there any chance of ending up with two observations with the same `new_depth` and `depth_diff`? Lines 190-191 would weed out identical original depths but wouldn't catch it if those two observations were at, say, 0.4 and 0.6m. If there were such duplicates, I think you probably would notice when writing out the observations in the date x depth matrix format, at least if using `pivot_wider`...but anyway, even recognizing all that, it might still be a good idea to ensure we get a single unique value here. You could do it by adding a call to `slice(1)` right after this filter call and before the `ungroup`, I think.

PS - here's why I think you'd notice when using pivot_wider: it puts a list in each cell if there are any duplicates and you don't specify `values_fn`:
```r
> tibble(depth=c(1,2,3,3,3,4,4,5), temp=1:8) %>% pivot_wider(names_from=depth, values_from=temp)
# A tibble: 1 x 5
  `1`       `2`       `3`       `4`       `5`      
  <list>    <list>    <list>    <list>    <list>   
1 <int [1]> <int [1]> <int [3]> <int [2]> <int [1]>
Warning message:
Values are not uniquely identified; output will contain list-cols.
* Use `values_fn = list` to suppress this warning.
* Use `values_fn = length` to identify where the duplicates arise
* Use `values_fn = {summary_fun}` to summarise duplicates 
> tibble(depth=c(1,2,3,3,3,4,4,5), temp=1:8) %>% pivot_wider(names_from=depth, values_from=temp, values_fn=mean)
# A tibble: 1 x 5
    `1`   `2`   `3`   `4`   `5`
  <dbl> <dbl> <dbl> <dbl> <dbl>
1     1     2     4   6.5     8
```